Configuration File Update w/o Certificate Issuance

ohemorange opened the pull request:
Fixes #8822

This PR allows a user to modify the configuration of a certificate without renewing it. This can be used to add or modify hooks, change the authenticator, or modify other renewal options. It will attempt a dry run in the process to verify that the new options work. 

Also creates a `reconfigure` help section:
```
$ certbot help reconfigure
usage: 

  certbot reconfigure --cert-name CERTNAME [options]

optional arguments:
  -h, --help            show this help message and exit
  -c CONFIG_FILE, --config CONFIG_FILE
                        path to config file (default: /etc/letsencrypt/cli.ini and ~/.config/letsencrypt/cli.ini)

reconfigure:
  Common options that may be updated with the "reconfigure" subcommand:

  --cert-name CERTNAME  Certificate name to apply. This name is used by Certbot for housekeeping and in file paths; it doesn't
                        affect the content of the certificate itself. To see certificate names, run 'certbot certificates'. When
                        creating a new certificate, specifies the new certificate's name. (default: the first provided domain or
                        the name of an existing certificate on your system for the same domains)
  --run-deploy-hooks    When performing a test run using `--dry-run` or `reconfigure`, run any applicable deploy hooks. This
                        includes hooks set on the command line, saved in the certificate's renewal configuration file, or present
                        in the renewal-hooks directory. To exclude direcory hooks, use --no-directory-hooks. The hook(s) will only
                        be run if the dry run succeeds, and will use the current active certificate, not the temporary test
                        certificate acquired during the dry run. This flag is recommended when modifying the deploy hook using
                        `reconfigure`. (default: False)
  --pre-hook PRE_HOOK   Command to be run in a shell before obtaining any certificates. Intended primarily for renewal, where it
                        can be used to temporarily shut down a webserver that might conflict with the standalone plugin. This will
                        only be called if a certificate is actually to be obtained/renewed. When renewing several certificates
                        that have identical pre-hooks, only the first will be executed. (default: None)
  --post-hook POST_HOOK
                        Command to be run in a shell after attempting to obtain/renew certificates. Can be used to deploy renewed
                        certificates, or to restart any servers that were stopped by --pre-hook. This is only run if an attempt
                        was made to obtain/renew a certificate. If multiple renewed certificates have identical post-hooks, only
                        one will be run. (default: None)
  --deploy-hook DEPLOY_HOOK
                        Command to be run in a shell once for each successfully issued certificate. For this command, the shell
                        variable $RENEWED_LINEAGE will point to the config live subdirectory (for example,
                        "/etc/letsencrypt/live/example.com") containing the new certificates and keys; the shell variable
                        $RENEWED_DOMAINS will contain a space-delimited list of renewed certificate domains (for example,
                        "example.com www.example.com") (default: None)
  -a AUTHENTICATOR, --authenticator AUTHENTICATOR
                        Authenticator plugin name. (default: None)
  -i INSTALLER, --installer INSTALLER
                        Installer plugin name (also used to find domains). (default: None)
  --webroot             Obtain certificates by placing files in a webroot directory. (default: False)
```

alexzorin reviewed (commented):
Yaaay. This is really cool and needed.

I haven't dug into the code yet but I've played with it a bit and it seems to work really well. Some high level feedback below.

ohemorange commented on certbot/certbot/_internal/main.py:
```diff
@@ -1652,6 +1653,76 @@ def make_or_verify_needed_dirs(config: configuration.NamespaceConfig) -> None:
+    _get_and_save_cert(le_client, lineage_config, certname=certname,
+        lineage=renewal_candidate)
+
+    # this function will update lineage.configuration with the new values, and save it to disk
+    renewal_candidate.save_new_config_values(lineage_config)
```
Tracking what was actually updated as opposed to what was there previously...not too bad code-wise, just violates a couple layers of interfaces, which isn't ideal. I can take a stab at it to see how bad it is. Definitely pro printing a success/failure message though, great call.

ohemorange replied:
So turns out this is reasonable, as long as we're ok with printing the internal names. Getting the argparse flags back out from the destination is tricky, especially because multiple flags can map onto one destination. I can look into that, but how firmly are you against just printing the internal name? For example, `--nginx` will tell you that `authenticator` and `installer` were updated, but like, that's probably understandable.

ohemorange replied:
Currently looks like:

```
$ certbot reconfigure --cert-name example.com --apache --pre-hook "echo new pre hook"
Saving debug log to /var/log/letsencrypt/letsencrypt.log
Hook 'pre-hook' ran with output:
 new pre hook
Simulating renewal of an existing certificate for example.com

 Successfully updated configuration.
 The following options were added:
    pre_hook: echo new pre hook
 The following options were changed:
    installer: nginx --> apache
    authenticator: nginx --> apache

$ certbot reconfigure --cert-name example.com --apache --pre-hook "echo new pre hook"
Saving debug log to /var/log/letsencrypt/letsencrypt.log
Hook 'pre-hook' ran with output:
 new pre hook
Simulating renewal of an existing certificate for example.com

 No changes were made to the renewal configuration.
```

alexzorin replied:
I think if a straightforward reverse mapping isn't doable, it's better to nix this. Exposing the internal flags in this manner seems antithetical to all the work we did on the UX previously.

Otherwise, I think the output looks great here.

ohemorange replied:
Even if there is one for the direct command line options, which I don't think there is, we'd still have to manually add stuff for configurators and such; we're manually doing things like separating `--nginx` into `-i nginx -a nginx`. I only looked briefly though, so I might have missed it in the argparse docs. But then you can still use either `deploy-hook` or `renew-hook` to set `deploy_hook` I believe, so we'd have to go in and manually specify there as well.

The only other thought I have there is just printing the values and not the keys, but that might make things more confusing.

alexzorin replied:
OK, let's cut it.

ohemorange replied:
Cut. New output:

```
$ certbot reconfigure --cert-name example.com --nginx
Saving debug log to /var/log/letsencrypt/letsencrypt.log
Hook 'pre-hook' ran with output:
 new pre hook
Simulating renewal of an existing certificate for example.com
Hook 'post-hook' ran with output:
 newpost

No changes were made to the renewal configuration.
$ # stop nginx
$ certbot reconfigure --cert-name example.com --apache
Saving debug log to /var/log/letsencrypt/letsencrypt.log
Hook 'pre-hook' ran with output:
 new pre hook
Simulating renewal of an existing certificate for example.com
Hook 'post-hook' ran with output:
 newpost

Successfully updated configuration.
Changes will apply when the certificate renews.
```

ohemorange commented on certbot/certbot/_internal/main.py:
```diff
@@ -1652,6 +1653,76 @@ def make_or_verify_needed_dirs(config: configuration.NamespaceConfig) -> None:
+    :raises errors.ConfigurationError: if certificate could not be loaded
+
+    """
+
```
I think "changes apply at the next renewal" is the way to go. I assume the distinction you were thinking was "these will apply to the cert itself" vs. "these apply to the webserver/local configuration"? Even if it applies to the cert itself, I can see it being useful to decouple changing what you want it to do and actually doing it in cases where rate limits might apply or some such.

osirisinferi replied:
Would a renewal, ahead of usual renewal time window, with updated `--key-type` et cetera require a `--force-renewal`? Or does Certbot notice differences between the current cert and `reconfigure`'d renewal param conf file, so that `--force-renewal` isn't necessary? I think it would be nice if Certbot noticed.

ohemorange replied:
It currently just checks ocsp (maybe) and cert expiry date, `--force-renewal` is necessary.

I currently added `Changes will apply when the certificate renews.` if anything was changed. Debating mentioning `--force-renewal` here, but leanings towards not doing so.

I agree noticing those differences would be nice to have, but I think it's not the simplest thing to do unless again we start listing particular terms we want here.

ohemorange commented on certbot/certbot/_internal/main.py:
```diff
@@ -1652,6 +1653,76 @@ def make_or_verify_needed_dirs(config: configuration.NamespaceConfig) -> None:
         util.make_or_verify_dir(hook_dir, strict=config.strict_permissions)
 
 
+def reconfigure(config: configuration.NamespaceConfig,
```
Or pass with "no changes have been made"? Either way definitely needs its own message.

ohemorange replied:
It is much easier to test at the end of everything if we've changed anything at all, which is a superset of just setting the certname; output in the comment above.

osirisinferi commented on certbot/certbot/_internal/main.py:
```diff
@@ -1652,6 +1653,76 @@ def make_or_verify_needed_dirs(config: configuration.NamespaceConfig) -> None:
+        raise errors.ConfigurationError("You have specified domains, but this function cannot "
+            "be used to modify the domains in a certificate. If you would like to do so, you "
+            "will need to renew the certificate instead of reconfiguring, which can be done "
+            "with the flag --force-renewal. Otherwise, remove the domains from the command "
```
Is `--force-renewal` actually necessary? I would recommend to mention `--force-renewal` as little as possible, as users on the Community often don't understand its meaning, purpose and potential rate limit hazard.

That said, the `renew` subcommand also doesn't work with `-d` options ("Currently, the renew verb is capable of either renewing all installed certificates that are due to be renewed or renewing a single certificate specified by its name. If you would like to renew specific certificates by their domains, use the certonly command instead. The renew verb may provide other options for selecting certificates to renew in the future.")

Users are required to use the `certonly` subcommand and with that, Certbot will ask to *update* the cert, even without `--force-renewal`.

ohemorange replied:
I mean, kind of. `--force-renewal` is just the flag to bypass the "are you sure you want to do this" question, so those are essentially the same thing, just one you tell it interactively and one you tell it on the command line. We can certainly say "which can be done by running `certbot certonly` instead of `certbot reconfigure`" and let that all happen interactively though, if we want to minimize references to `--force-renewal`.

alexzorin replied:
I agree with Osiris here in terms of this phrase:

>you will need to renew the certificate instead

The `renew` verb is not applicable here and it's not really a renew operation. It's going to be `certonly` or `run` with `--cert-name` set and `--force-renewal`. Quite a mouthful.

My proposal: create a new certbot.org short link and link to https://eff-certbot.readthedocs.io/en/stable/using.html#changing-a-certificate-s-domains from this error message. And make sure that section doesn't get wiped out by Alexis' work in #9439.

alexzorin replied:
I still think we should change the language here.

To say "you will need to renew the certificate instead" while our `renew` verb is not actually capable of changing domains on the certificate, will end up being a wild goose chase for the user.

ohemorange commented on certbot/certbot/_internal/main.py:
```diff
@@ -1652,6 +1653,76 @@ def make_or_verify_needed_dirs(config: configuration.NamespaceConfig) -> None:
         util.make_or_verify_dir(hook_dir, strict=config.strict_permissions)
 
 
+def reconfigure(config: configuration.NamespaceConfig,
```
Given the rewrite going on in https://github.com/certbot/certbot/pull/9439, I think let's get this merged first and then figure out what to document, but I totally agree, including that as an example would be great.

ohemorange replied:
This does actually remind me though, since it's currently just doing a dry run, it's not actually running the deploy hook. Since you can modify the deploy hook, it should probably test that as well. Just...need to figure out how.

ohemorange replied:
Ok, did this! Only mildly sketchy.

ohemorange replied:
```
$ certbot reconfigure --cert-name example.com --deploy-hook 'echo $RENEWED_LINEAGE '
Saving debug log to /var/log/letsencrypt/letsencrypt.log
Simulating renewal of an existing certificate for example.com
Hook 'deploy-hook' ran with output:
 /etc/letsencrypt/live/one.funkycat.space

Successfully updated configuration.
The following options were changed:
    renew_hook: cp $RENEWED_LINEAGE /tmp/testcb/ --> echo $RENEWED_LINEAGE 
Changes will apply when the certificate renews.
```

ohemorange replied:
Which means this would now fix https://github.com/certbot/certbot/issues/5658 as well

alexzorin replied:
I have a couple of thoughts about `reconfigure` running the deploy hook:

1. I am not crazy about a dry-run doing a wet operation as the default. Some deploy hooks in the wild, I think, are going to attract some big costs in terms of downtime or resources or rate limits or whatever.
2. I think it should match the behavior of `--dry-run`, since it is a superset of that operation, and users may form that understanding of the verb.
3. I think it is a valuable behavior to have and that #5658 is worth fixing. Therefore, I propose the solution of a non-default flag that will affect both `renew --dry-run` and `reconfigure`. And for `reconfigure`, to possibly emit a warning/suggestion if a `--deploy-hook` is added/changed but the flag is omitted.

ohemorange replied:
Hm. Why is it ok to call pre- and post- hooks on dry-run but not deploy? Because deploy tends to more commonly involve calling out to other servers, whereas pre- and post- tend to be local? Theoretically deploy should be idempotent and actually deploy the last live information, not whatever we're touching in this dry run.

As for matching `--dry-run`, thinking about it, it's kind of weird that `certbot renew --dry-run` doesn't run the deploy hook in the first place. I suspect that was less of an active decision and more of a side effect of the implementation path that `--dry-run` uses. I do agree about them matching though, and likewise think your suggestion is a good way to handle a mismatch in behavior.

alexzorin replied:
>Why is it ok to call pre- and post- hooks on dry-run but not deploy?

Practically speaking, pre- and post- hooks might be required for the authenticator to succeed: starting/stopping servers, changing firewall rules. In that case, we must run them for a representative test.

For deploy hooks, I see your point. We could change it to be the default behavior for `--dry-run`. I can see it going either way, if you feel strongly about doing it by default. I'd be more comfortable if there was a flag to enable/disable the behavior though, whatever decision to make. And that flag would have to make sense in conjunction with `--no-directory-hooks`.

ohemorange replied:
Thinking about it more, since the thing we're changing with this command is the renewal config, what we're doing here is a dry run of a renewal, as opposed to the only previously existing dry run command, which was a dry run of a certonly. In that case it makes sense to have the auth stuff only, but here I really do think it makes sense to have it do the deploy hook by default. Which is a little awkward, because I can absolutely see how people would expect dry run and this to have similar behavior. It's also possible that that expectation is created by having written/read the code here. If that's the case, we can just be careful how we talk about it, and leave the behavior as-is, along with adding a command like `--do-not-run-deploy-hooks`. Is that too messy of a behavior, do you think?

alexzorin replied:
>what we're doing here is a dry run of a renewal, as opposed to the only previously existing dry run command, which was a dry run of a certonly

So I'm *probably* misunderstanding your point, but do we not already have dry-runs of renewals?

    # certbot >>renew<< --dry-run
    Saving debug log to /var/log/letsencrypt/letsencrypt.log

    - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
    Processing /etc/letsencrypt/renewal/army-p-nutty.zorin.au.conf
    - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
    Simulating >>renewal<< of an existing certificate for army-p-nutty.zorin.au

`certbot renew --dry-run` and `certbot reconfigure` need to have the same behavior with respect to `--deploy-hook`, I think. The only difference between these commands is that `reconfigure` persists the changes, and I don't think that has an effect on whether the deploy hook runs or not.

I am still quite worried about running the deploy hook by default. I would much prefer to have the friction of a prompt/opt-in over a potentially unsafe (e.g. an expired certificate that the user is trying to fix) deploy.

ohemorange replied:
Hm, I did forget that you can do `renew --dry-run`, yes; too much looking at the code, where the dry run functionality is only in one place, the get-cert place.

Ok, what if we detect if the user is trying to change/set the deploy hook that lives in the renewal config (that is, just look for `--deploy-hook`, don't look at directory hooks), and in that case prompt to opt-in (along with flag to bypass prompt)? Thinking of this as a compromise, because it feels quite weird to let users change the setting without testing it at all in which case it will just fail next renewal without the user there, but at the same time, don't want to always bring up an annoying prompt every time they change anything.

Might also be nice to have that flag work for regular dry runs as well, outside of `reconfigure`.

alexzorin replied:
>what if we detect if the user is trying to change/set the deploy hook that lives in the renewal config (that is, just look for --deploy-hook, don't look at directory hooks), and in that case prompt to opt-in (along with flag to bypass prompt)?

>Might also be nice to have that flag work for regular dry runs as well, outside of reconfigure.

Agreed on all points!

alexzorin commented on certbot/certbot/_internal/cli/verb_help.py:
```diff
@@ -102,6 +102,11 @@
         "usage": "\n\n  certbot show_account [options]\n\n"
     }),
+    ("reconfigure", {
+        "short": "Update configuration information for a certificate specified by --cert-name",
```
Do we call this "information" anywhere else? Maybe "renewal configuration" or just "configuration"?

```suggestion
        "short": "Update configuration for a certificate specified by --cert-name",
```

alexzorin commented on certbot/certbot/_internal/main.py:
```diff
@@ -1654,6 +1655,129 @@ def make_or_verify_needed_dirs(config: configuration.NamespaceConfig) -> None:
+
+    # figure this out before we modify config
+    if config.deploy_hook and not config.run_deploy_hook:
+        msg = "You are attempting to set a new deploy hook. Would you like Certbot to run the new "
+              "hook when it performs a dry run with the new settings? This will run all relevant "
+              "deploy hooks, including directory hooks, unless --no-directory-hooks is set."
```
Missing `(` `)` around multi-line string.

alexzorin commented on certbot/certbot/_internal/main.py:
```diff
@@ -1654,6 +1655,129 @@ def make_or_verify_needed_dirs(config: configuration.NamespaceConfig) -> None:
+
+    # figure this out before we modify config
+    if config.deploy_hook and not config.run_deploy_hook:
+        msg = "You are attempting to set a new deploy hook. Would you like Certbot to run the new "
```
Very minor thing, if you like it: I've been in the habit of always using the flag name where possible, to make things more obvious to the user.

```suggestion
        msg = "You are attempting to set a new --deploy-hook. Would you like Certbot to run the new "
```

ohemorange replied:
love this!

alexzorin commented on certbot/certbot/_internal/main.py:
```diff
@@ -1654,6 +1655,129 @@ def make_or_verify_needed_dirs(config: configuration.NamespaceConfig) -> None:
+    # figure this out before we modify config
+    if config.deploy_hook and not config.run_deploy_hook:
+        msg = "You are attempting to set a new deploy hook. Would you like Certbot to run the new "
+              "hook when it performs a dry run with the new settings? This will run all relevant "
```
Do we need to do any explaining around which certificate the deploy hook is being invoked against? i.e. that this is semantically a dry-run and that we didn't persist  the test certificate to disk?

I'm not sure at the moment how obvious all of that is to the user, at the moment.

ohemorange replied:
Oh, that's the great point, I'll add that.

alexzorin commented on certbot/certbot/_internal/cli/__init__.py:
```diff
@@ -162,6 +162,16 @@ def prepare_and_parse_args(plugins: plugins_disco.PluginsRegistry, args: List[st
+        ["testing", "renew", "certonly", "reconfigure"],
+        "--run-deploy-hook", action="store_true", dest="run_deploy_hook",
+        default=flag_default("run_deploy_hook"),
+        help="When performing a test run using `--dry-run` or `reconfigure`, run any applicable"
```
As with the other comment, do we need to mention that the deploy hook will be running against the most recent live certificate?

alexzorin commented on certbot/certbot/_internal/main.py:
```diff
@@ -1654,6 +1655,129 @@ def make_or_verify_needed_dirs(config: configuration.NamespaceConfig) -> None:
+    orig_renewal_params = orig_renewal_conf['renewalparams']
+    orig_renewal_params_set = set(orig_renewal_params.items())
+    final_renewal_params = final_renewal_conf['renewalparams']
+    final_renewal_params_set = set(final_renewal_params.items())
```
I have a crash here, after the changes get persisted to disk.

Presumably it has something to do with `--webroot-path` being a `Union[str, Sequence[Any], None]`.

    $ certbot certonly --standalone --http-01-port 8123 -d example.com
    $ certbot reconfigure --cert-name example.com --webroot -w /tmp

    Traceback (most recent call last):
      File "/Users/alex/devel/certbot/venv/bin/certbot", line 33, in <module>
        sys.exit(load_entry_point('certbot', 'console_scripts', 'certbot')())
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
      File "/Users/alex/devel/certbot/certbot/certbot/main.py", line 19, in main
        return internal_main.main(cli_args)
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
      File "/Users/alex/devel/certbot/certbot/certbot/_internal/main.py", line 1861, in main
        return config.func(config, plugins)
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
      File "/Users/alex/devel/certbot/certbot/certbot/_internal/main.py", line 1779, in reconfigure
        _report_reconfigure_results(renewal_file, orig_renewal_conf)
      File "/Users/alex/devel/certbot/certbot/certbot/_internal/main.py", line 1681, in _report_reconfigure_results
        final_renewal_params_set = set(final_renewal_params.items())
                                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    TypeError: unhashable type: 'list'  TypeError: unhashable type: 'list'

It would be good to get a test for this, and probably an integration test would be good too!

ohemorange replied:
lol yes thanks for catching this in the early review, but yes, I still planned to add tests. Don't know that I would have thought to check this particular one or that I expected it to not Just Work there, so helpful nonetheless!

alexzorin reviewed (changes requested):
I wasn't sure if you were ready for further review but here it is :D.

alexzorin commented on certbot/certbot/_internal/main.py:
```diff
@@ -1654,6 +1655,128 @@ def make_or_verify_needed_dirs(config: configuration.NamespaceConfig) -> None:
+            "be found. Run `certbot certificates` to list available certificates.")
+
+    # figure this out before we modify config
+    if config.deploy_hook and not config.run_deploy_hook:
+        msg = ("You are attempting to set a new --deploy-hook. Would you like Certbot to run the "
```
If the new `--deploy-hook` set on the CLI, is the same as the old `--deploy-hook` set in the renewal configuration, the message is not quite true.

What do you think about just cutting the word "new" out? That way we can prompt the user to test the hook, even if the contents of the hook have not changed.

ohemorange replied:
I mean, technically true, it's just that the new one is the same as the old one, but I see how it can be confusing. I'll cut the `new`.

alexzorin commented on certbot/certbot/_internal/cli/__init__.py:
```diff
@@ -162,6 +162,17 @@ def prepare_and_parse_args(plugins: plugins_disco.PluginsRegistry, args: List[st
              " renewal. --deploy-hook commands are not called.")
+    helpful.add(
+        ["testing", "renew", "certonly", "reconfigure"],
+        "--run-deploy-hook", action="store_true", dest="run_deploy_hook",
+        default=flag_default("run_deploy_hook"),
```
Should "hook" here be pluralized?

```suggestion
        "--run-deploy-hooks", action="store_true", dest="run_deploy_hooks",
        default=flag_default("run_deploy_hooks"),
```

ohemorange replied:
I went back and forth on this one -- my thinking here was that the most common usage would be when it's set to test for a newly set-on-the-cli deploy hook, in which case we're just talking about the one. But pluralizing would make it more clear that it will in fact run any and all present ones, so that's probably for the best, now that I've gone and written the explanatory text.

alexzorin reviewed (commented):
The latest set of changes look good!

I think we'll need to stick `# pylint: disable=too-many-lines` at the top of `storage.py`.

alexzorin reviewed (commented):
Besides the [error message about modifying domains on the certificate](https://github.com/certbot/certbot/pull/9355#discussion_r1045181316), this feels pretty finished!

I had suggested two things earlier:

1. An integration test that uses the `reconfigure` verb, just so we remain really sure that it really works.
2. Documenting the usage of this verb in the user guide. (I am 100% behind punting this to a follow-up PR though).

Also, master will need to be merged for CHANGELOG.

ohemorange commented on certbot/certbot/_internal/main.py:
```diff
@@ -1654,6 +1655,127 @@ def make_or_verify_needed_dirs(config: configuration.NamespaceConfig) -> None:
+    if config.domains:
+        raise errors.ConfigurationError("You have specified domains, but this function cannot "
+            "be used to modify the domains in a certificate. If you would like to do so, follow "
+            "the instructions at https://certbot.org/change-cert-domain. Otherwise, remove the "
```
link does not yet exist, do not merge!

alexzorin replied:
The link seems to have appeared today, let's merge! 🎉

ohemorange commented:
> Besides the error message about modifying domains on the certificate, this feels pretty finished!

Updated the message; waiting for shortlink to be created. I don't think it will interfere with any documentation changes, but we can always update the link location if needed.

> An integration test that uses the reconfigure verb, just so we remain really sure that it really works.

Added a simple test for adding and updating, though I kept it basic since the unit tests are in fact updating the configuration file as well. 
 
> Documenting the usage of this verb in the user guide. (I am 100% behind punting this to a follow-up PR though).

Created https://github.com/certbot/certbot/issues/9531

> Also, master will need to be merged for CHANGELOG.

Done

alexzorin reviewed (dismissed):
LGTM!
